drivers/sensor/mpu9250/ak8963.c

```diff
@@ -0,0 +1,348 @@
+	 * given device register with given details
+	 */
+	const struct mpu9250_config *cfg = dev->config;
+	uint8_t mode_bit = 0x00;
+	uint8_t status;
+	int ret;
+
+	if (write == false) {
+		mode_bit = I2C_READ_FLAG;
+	}
+
+	/* Set target i2c address */
+	ret = i2c_reg_write_byte_dt(&cfg->i2c,
+				    MPU9250_REG_I2C_SLV4_ADDR,
+				    AK8963_I2C_ADDR | mode_bit);
+	if (ret < 0) {
+		LOG_ERR("Failed to write i2c target slave address.");
+		return ret;
+	}
+
+	/* Set target i2c register */
+	ret = i2c_reg_write_byte_dt(&cfg->i2c,
+				    MPU9250_REG_I2C_SLV4_REG,
+				    reg);
+	if (ret < 0) {
+		LOG_ERR("Failed to write i2c target slave register.");
+		return ret;
+	}
+
+	/* Initiate transfer  */
+	ret = i2c_reg_write_byte_dt(&cfg->i2c,
+				    MPU9250_REG_I2C_SLV4_CTRL,
+				    MPU9250_REG_I2C_SLV4_CTRL_VAL);
+	if (ret < 0) {
+		LOG_ERR("Failed to initiate i2c slave transfer.");
+		return ret;
+	}
+
+	/* Wait for a transfer to be ready */
+	while (!(status & MPU9250_I2C_MST_STS_SLV4_DONE)) {
+		ret = i2c_reg_read_byte_dt(&cfg->i2c,
+					   MPU9250_I2C_MST_STS, &status);
+		if (ret < 0) {
+			LOG_ERR("Waiting for slave failed.");
+			return ret;
+		}
+
+		k_msleep(10);
+	}
+
+	return 0;
+}
+
+static int ak8963_read_reg(const struct device *dev, uint8_t reg, uint8_t *data)
+{
+	const struct mpu9250_config *cfg = dev->config;
+	int ret;
+
+	/* Execute transfer */
+	ret = ak8963_execute_rw(dev, reg, false);
+	if (ret < 0) {
+		LOG_ERR("Failed to prepare transfer.");
+		return ret;
+	}
+
+	/* Read the result */
+	ret = i2c_reg_read_byte_dt(&cfg->i2c,
+				   MPU9250_REG_I2C_SLV4_DI, data);
+	if (ret < 0) {
+		LOG_ERR("Failed to read data from slave.");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int ak8963_write_reg(const struct device *dev, uint8_t reg, uint8_t data)
+{
+	const struct mpu9250_config *cfg = dev->config;
+	int ret;
+
+	/* Set the data to write */
+	ret = i2c_reg_write_byte_dt(&cfg->i2c,
+				    MPU9250_REG_I2C_SLV4_DO, data);
+	if (ret < 0) {
+		LOG_ERR("Failed to write data to slave.");
+		return ret;
+	}
+
+	/* Execute transfer */
+	ret = ak8963_execute_rw(dev, reg, true);
+	if (ret < 0) {
+		LOG_ERR("Failed to transfer write to slave.");
+		return ret;
+	}
+
+	return 0;
+}
+
+
+static int ak8963_set_mode(const struct device *dev, uint8_t mode)
+{
+	int ret;
+
+	ret = ak8963_write_reg(dev, AK8963_REG_CNTL1, mode);
+	if (ret < 0) {
+		LOG_ERR("Failed to set AK8963 mode.");
+		return ret;
+	}
+
+	/* Wait for mode to change */
+	k_msleep(1);
+	return 0;
+}
+
+static int16_t ak8963_calc_adj(int16_t val)
+{
+
+	/** Datasheet says the actual register value is in 16bit output max
+	 *  value of 32760 that corresponds to 4912 uT flux, yielding factor
+	 *  of 0.149938.
+	 *
+	 *  Now Zephyr unit is Gauss, and conversion is 1T = 10^4G
+	 *  -> 0.1499 * 10^4 = 1499
+	 *  So if we multiply with scaling with 1499 the unit is uG.
+	 */
+	return ((AK9863_SCALE_TO_UG * (val - 128)) / 256) + AK9863_SCALE_TO_UG;
```

Comment:
Could you add an explanation of the calculations done here?

Reply:
Added original from docs.